### DISCUSSION TITLE
Complete directory arguments with "/" after menu completion?

### DISCUSSION THREAD

**Discussion by mozirilla213** [akinomyoga/ble.sh#297] [Q&A] [answered]
When there's a single option for completion ble adds a trailing slash to directories:
```bash
$ cd /e<TAB>
$ cd /etc/
```
But if there are multiple choices it doesn't:
```bash
$ cd /s<TAB><TAB><TAB>
/sbin@ /srv/ /sys/
$ cd /srv
```
`bind 'set mark-directories on'` only makes trailing slashes to show, but not insert them when scrolling through options. Is there a way to enable this behaviour?

By the way, is there a way that the first `<TAB>` completes the first choice? right now `cd /s<TAB>` shows the choices and an autosuggestion to the first choice, but I'd like for it to complete the first choice on the command line by default, but still show the other choices and being able to `<TAB>` between them. I cannot seem to find a readline/ble option.

**Comment by akinomyoga** [comment, marked as answer]
A trailing slash is added when the completion is completed.

> When there's a single option for completion ble adds a trailing slash to directories:
> 
> ```shell
> $ cd /e<TAB>
> $ cd /etc/
> ```

This is because the completion immediately completes when there is a unique candidate.

> But if there are multiple choices it doesn't:
> 
> ```shell
> $ cd /s<TAB><TAB><TAB>
> /sbin@ /srv/ /sys/
> $ cd /srv
> ```

While selecting a completion candidate in the menu, the trailing slash isn't added but will later be added on the completion of the menu completion. You need to press <kbd>RET</kbd> to complete the completion, or otherwise the completion would be canceled prematurely.

> `bind 'set mark-directories on'` only makes trailing slashes to show,

That's `bind 'set visible-stats on'`. `mark-directories` is unrelated to the suffix shown in the menu.

> but not insert them when scrolling through options. Is there a way to enable this behaviour?

You can write your own programmable completion so that the candidates include the trailing slashes within themselves. (But as far as `mark-directories` is turned on, with ble.sh's built-in filename completions, the trailing slash is anyway added when completing the completion.)

----

> By the way, is there a way that the first `<TAB>` completes the first choice? right now `cd /s<TAB>` shows the choices and an autosuggestion to the first choice, but I'd like for it to complete the first choice on the command line by default, but still show the other choices and being able to `<TAB>` between them. I cannot seem to find a readline/ble option.

Have you tried `menu-complete`?

```bash
# Widget "menu-complete" in ble
ble-bind -f C-i menu-complete
ble-bind -f TAB menu-complete

# Bindable function "menu-complete" in readline (plain Bash without ble)
bind '"\C-i": menu-complete'
```

but maybe the behavior of the widget `menu-complete` for the unique completion might not be what you expect.
